### PR TITLE
Support tuple indexing in CuPy JIT

### PR DESCRIPTION
**Pull request by asi1024**
Depends on #4890

### CODE REVIEW

**Comment by asi1024** [comment]
@emcastillo Rebased. PTAL!

**Review by emcastillo — DISMISSED** [review]

File: cupyx/jit/_compile.py [hunk 1]
```diff
@@ -546,18 +546,47 @@ def _transpile_expr_internal(expr, env):
...
+                    f'{value.code}[{index.code}]',
+                    _types.Scalar(value.ctype.dtype))
+            if isinstance(index.ctype, _types.Tuple):
+                if ndim != len(index.ctype.types):
+                    raise IndexError(f'The size of index must be {ndim}')
+                for t in index.ctype.types:
+                    if not isinstance(t, _types.Scalar):
+                        raise TypeError('Array indices must be scalar.')
+                    if t.dtype.kind not in 'iu':
+                        raise TypeError('Array indices must be integer.')
```
Comment (emcastillo) [inline cupyx/jit/_compile.py:581]: I wonder if an empty tuple would hard error before arriving here.

**Comment by emcastillo** [comment]
Jenkins, test this please

File: cupyx/jit/_compile.py (hunk 1 above)
Comment (asi1024) [reply cupyx/jit/_compile.py:581]: Empty tuples will be caught here:
https://github.com/cupy/cupy/pull/4939/files#diff-92c76e1dfae3dd9e376fac22b852d0deb6cea6d0affa13bda3ac7d2ee33ce17dR575-R576
, but indexing of 0-dim array is not tested. I will add the test in this PR.

Comment (asi1024) [reply cupyx/jit/_compile.py:581]: I noticed 0-dim array does not have `_indexing` member function. Fixed in cbf02f3!

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**